### ISSUE TITLE
在使用和对vxeTable二次封装时，大量测试和实践后发现来一下功能缺陷和需要扩展的地方，

### ISSUE THREAD

**Issue by Titans1001** [x-extends/vxe-table#1001]
**（必填）这个需求解决了什么问题：**
在使用和对vxeTable二次封装时，大量测试和实践后发现来一下功能缺陷和需要扩展的地方。
以下列举了vxe-table源码级别功能缺陷，为源码级别缺陷，需要进一步商讨处理方案和执行改进优化。这些只是一些迫切需要解决的缺陷和功能完善，尚有一些其他需要完善和改进的地方，个人联系邮箱：2396757501@qq.com，欢迎联系并交流改进意见。

1，可否增加类似expandConfig的值为false，直接关闭此项功能，类似于toolbarConfig配置为false关闭toolbar这样，推荐全局的类似于配置的值都允许设为Boolean和Object,以此来支持直接关闭此项功能。

2，全局的配置和vxe方法里面可否全局加上类似于渲染器的第二个参数context，以此来获取嵌套组件的this,执行其他操作，比如footerMethod方法，推荐全局的vxeTabble调用的方法第二个参数都加入context={$grid,$Table,$Form},以此来支持其他操作。

3，formatter里面虽然提供了context,但是注册后后期并不会触发原有的方法回调，这个可否改为一个内存地址，直接调用。

4，可否增加表格合计行配置比如第一行或最后一行，亦或者footer-config增加放置在第一行的配置。

5， renderer渲染器可否增加异步（同步）渲染，目前是如果加上 async，变为async renderEdit(){},其不再渲染。

6，renderer函数里面支持jsx和h(),当前支持接收的返回值必须为数组，如果直接return h()而非return [h()]报错。

7,    form可否增加itemChange事件优化增加增量以及变量重渲染而非整个form的重渲染以此来支持联动。

8，可否所有的回调都增加代理模式去解决一些问题，形如proxyEvent("回调函数名",...[arguments])，当$emit("methods"，...[arguments])的同时$emit("proxyEvent","methods"，...[arguments])

9，校验rules里面虽然支持message值为函数，但是并没有提供可用参数。

10，列过滤存在某一列过滤已经选择或输入规则但未点击筛选，又操作来另外一列筛选，点击筛选后两列的筛选条件同时起作用进行筛选。

**建议的 API 是什么样的：**
需要更具上述缺陷进一步完善和制定相应api和
**是否已有其他不错的替代方案：**
否

**Comment by xuliangzhan** [comment]
1.❌展开列的开关是 column.type='expend'
2.✔️可以支持，但在数据方法中去操作是属于错误的做法，请谨慎操作不当进入死循环
3.❓描述不明确
4.❌不支持
5.❌renderer是 VNode 行为，不存在异步
6.❌renderer只接收数组类型 string[]、VNode[]
7.❓描述不明
8.❌不合理
9.❌仅用字符串或国际化
10✔️下版本修复
11.✔️可以支持
12.✔️这是工具类的bug，更新最新版本即可

**Comment by Titans1001** [comment]
全局的expandconfig可否增加bool值全局关闭不让展开



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: xuliangzhan <notifications@github.com&gt;
发送时间: 2020年8月16日 12:17
收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt;
抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt;
主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)

**Comment by Titans1001** [comment]
可否加微信交流，我最近一直在用vxetable



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: xuliangzhan <notifications@github.com&gt;
发送时间: 2020年8月16日 12:17
收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt;
抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt;
主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)

**Comment by Titans1001** [comment]
我微信是19993170150



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: xuliangzhan <notifications@github.com&gt;
发送时间: 2020年8月16日 12:17
收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt;
抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt;
主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)

**Comment by Titans1001** [comment]
第五个之所以需要支持异步，是要借助await做联动远程数据拉取行联动



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: xuliangzhan <notifications@github.com&gt;
发送时间: 2020年8月16日 12:17
收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt;
抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt;
主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)





 
1.❌展开列的开关是 column.type='expend'
 2.✔️可以支持，但在数据方法中去操作是属于错误的做法，请谨慎操作不当进入死循环
 3.❓描述不明确
 4.❌不支持
 5.❌renderer是 VNode 行为，不存在异步
 6.❌renderer只接收数组类型 string[]、VNode[]
 7.❓描述不明
 8.❌不合理
 9.❌仅用字符串或国际化
 10✔️下版本修复
 11.✔️可以支持
 12.✔️这是工具类的bug，更新最新版本即可

**Comment by Titans1001** [comment]
第一个expandconfig是为了二次封装关闭或开启展开配置，如果传入空对象存在其他问题，所以可否支持传入false关闭，true，按默认配置走，对象，传入部分配置



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: xuliangzhan <notifications@github.com&gt;
发送时间: 2020年8月16日 12:17
收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt;
抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt;
主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)





 
1.❌展开列的开关是 column.type='expend'
 2.✔️可以支持，但在数据方法中去操作是属于错误的做法，请谨慎操作不当进入死循环
 3.❓描述不明确
 4.❌不支持
 5.❌renderer是 VNode 行为，不存在异步
 6.❌renderer只接收数组类型 string[]、VNode[]
 7.❓描述不明
 8.❌不合理
 9.❌仅用字符串或国际化
 10✔️下版本修复
 11.✔️可以支持
 12.✔️这是工具类的bug，更新最新版本即可

**Comment by xuliangzhan** [comment]
> 第五个之所以需要支持异步，是要借助await做联动远程数据拉取行联动 发自我的iPhone
> […](#)
> ------------------ 原始邮件 ------------------ 发件人: xuliangzhan <notifications@github.com&gt; 发送时间: 2020年8月16日 12:17 收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt; 抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt; 主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001) 1.❌展开列的开关是 column.type='expend' 2.✔️可以支持，但在数据方法中去操作是属于错误的做法，请谨慎操作不当进入死循环 3.❓描述不明确 4.❌不支持 5.❌renderer是 VNode 行为，不存在异步 6.❌renderer只接收数组类型 string[]、VNode[] 7.❓描述不明 8.❌不合理 9.❌仅用字符串或国际化 10✔️下版本修复 11.✔️可以支持 12.✔️这是工具类的bug，更新最新版本即可 — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

异步只得是数据异步，VNode 不存在异步，VNode 对自动对数据的变化去做相应的渲染

**Comment by xuliangzhan** [comment]
> 第一个expandconfig是为了二次封装关闭或开启展开配置，如果传入空对象存在其他问题，所以可否支持传入false关闭，true，按默认配置走，对象，传入部分配置 发自我的iPhone
> […](#)
> ------------------ 原始邮件 ------------------ 发件人: xuliangzhan <notifications@github.com&gt; 发送时间: 2020年8月16日 12:17 收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt; 抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt; 主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001) 1.❌展开列的开关是 column.type='expend' 2.✔️可以支持，但在数据方法中去操作是属于错误的做法，请谨慎操作不当进入死循环 3.❓描述不明确 4.❌不支持 5.❌renderer是 VNode 行为，不存在异步 6.❌renderer只接收数组类型 string[]、VNode[] 7.❓描述不明 8.❌不合理 9.❌仅用字符串或国际化 10✔️下版本修复 11.✔️可以支持 12.✔️这是工具类的bug，更新最新版本即可 — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

展开列的开关是 column.type='expend'，和expandconfig没关系，expandconfig 只用于设置参数，这和 toolbar 行为不一样

**Comment by Titans1001** [comment]
嗯呢,现在是加了async就不渲染了



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: xuliangzhan <notifications@github.com&gt;
发送时间: 2020年8月16日 12:36
收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt;
抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt;
主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)

**Comment by Titans1001** [comment]
加个微信哈，交流哈，19993170150，每周我们交流一次



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: xuliangzhan <notifications@github.com&gt;
发送时间: 2020年8月16日 12:36
收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt;
抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt;
主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)

**Comment by xuliangzhan** [comment]
> 嗯呢,现在是加了async就不渲染了 发自我的iPhone
> […](#)
> ------------------ 原始邮件 ------------------ 发件人: xuliangzhan <notifications@github.com&gt; 发送时间: 2020年8月16日 12:36 收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt; 抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt; 主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)

renderer 不存在异步，不能使用任何异步函数

**Comment by Titans1001** [comment]
加个微信哈，交流哈，19993170150，每周我们交流一次



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "逍遥子泰坦"                                                                                    <2396757591@qq.com&gt;;
发送时间:&nbsp;2020年8月16日(星期天) 中午12:40
收件人:&nbsp;"reply+ALAKI5IHCZVWXIDT22SRK6V5ISMMVEVBNHHCQYV7EQ"<reply@reply.github.com&gt;;

主题:&nbsp;回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)



加个微信哈，交流哈，19993170150，每周我们交流一次



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: xuliangzhan <notifications@github.com&gt;
发送时间: 2020年8月16日 12:36
收件人: x-extends/vxe-table <vxe-table@noreply.github.com&gt;
抄送: roninTitans <2396757591@qq.com&gt;, Author <author@noreply.github.com&gt;
主题: 回复：[x-extends/vxe-table] 在使用和对vxeTable二次封装时，大量测试和实践后发现了以下功能缺陷和需要扩展的地方 (#1001)

**Comment by Titans1001** [comment]
4，可以改变为可否增加一个展示在head下方表格列表上方的类似于footerConfig的body-custom-config:{
  showPosition:"top",  // 或"bottom"
  body-custom-method({columns, data, $grid, $table}){
    // 类似于footer渲染的方法在表格肢体列表上方会下方渲染展示一个统计或其他信息，因为现实中有这样的需求，这个对于你来说很简单的，但是对于我要支持这样的需求则需要改源码，每次更新后维护加入我改动的地方。比如列举一个简单的例子技术分享记录：有这样一个表格，表格header是每天的时间，第一行为每天的技术分享人名字和统计的参与人数，主体内容为参与技术讲座的参与记录，表尾部为每天的人数统计。主体内容列表肯定是个一块，但是实际需求中有可能在header下边主体列表上方显示其他信息。之所以提这个需求意义在于此。
  }
}

7，看了下面代码应该可以理解我想表达的意思
  $vxeFormSelect: {
    // 下拉选择联动触发
    renderItem(h, { props, name, options }, { $form, property, data }) {
        return [
          h('vxe-select', {
            props: {
              ref: 'vxe-select-' + property,
              value: data[property],
              ...props
            },
            on: {
              change: ({ value }) => {
                data[property] = value
                $form.$parent.itemChange({ $form, property, itemValue: value, data }); // 告诉父组件当前change项做联动操作，我所说的itemChange({ $form, property, itemValue: value, data })意义在于此,因为如果只使用vxe-form是一包数据，这种操作不支持。
              }
            }
          }, options.map(item => {
            return h('vxe-option', {
              props: {
                value: item.value,
                label: item.label
              }
            });
          })
          )
        ]
    }
  },
8的意义在于因为$emit是单向通知触发一个回调，代理模式我可以不用去逐个绑定回调，只需要一个全局代理就可以，需要时则在代理中增加一个策略方法，回调函数名做key，具体的方法做值，我只需要维护这个策略方法就OK。
绑定：@proxyEvent="proxyEvent"
 proxyEvent() {
  //   注册事件代理
  let arg=[...arguments]
  this['gloable'+arg.shift()](...arg)        
 }
export const tableEvent={
  // 只需要这个里面添加对应的方法就OK
   gloableCellClick(argObj){
    
   }
}
9,校验rules里面虽然支持message值为函数，但是并没有提供可用参数,我的意思是可以借助提供的参数动态返回一个精确的提示，比如当一个复杂的校验需要满足很多条件时，我就可以精确的根据结果动态的引入一个全局配置库，然后根据配置库对应的文字来显示校验提示，规则也是一样。这样后期如果改变规则我只需要改变配置库就可以。

**Comment by Titans1001** [comment]
分页存在bug,即使配置了pageSizes:[20,50,100,1000,10000,100000],但是首次加载分页回调传回的pageSize任然为10。

**Comment by sushuier** [comment]
关于第5点，不知道我的需求是否正确，我补充说明一个情况：
后台返回的表格数据，某个字段 status，返回的值可能是1 或 0，需要在表格中显示为 正常 或 异常，自定义渲染时，要先获取状态字典，然后再根据 status渲染中文字符串，就会涉及异步渲染问题了；vnode的确不存在任何异步行为，但业务实践过程中还是会有这方面需求；

如果在 gridOptions.proxyConfig.ajax.query 查询表格数据之前 await 调用字典数据，然后写到list的返回数据中，那么每次请求列表都会重复请求字典；实际情况是：字典会在页面加载完成后只请求一次，然后每次切换列表分页或查询列表操作时，在列表返回值中完成字典映射；

不知道 我这个需求 当前插件有没有办法实现；
